api.go

```diff
@@ -24,32 +24,42 @@ import (
 
 const tokenEndpoint = "/accesstoken"
 
-func InitAPI(services apid.Services) {
-	services.API().HandleFunc(tokenEndpoint, getAccessToken).Methods("GET")
+const (
+	parBlock = "block"
+	parTag   = "If-None-Match"
+)
+
+type ApiManager struct {
+	tokenMan tokenManager
+	endpoint string
 }
 
-func getAccessToken(w http.ResponseWriter, r *http.Request) {
-	b := r.URL.Query().Get("block")
+func (a *ApiManager) InitAPI(api apid.APIService) {
+	api.HandleFunc(a.endpoint, a.getAccessToken).Methods("GET")
+}
+
+func (a *ApiManager) getAccessToken(w http.ResponseWriter, r *http.Request) {
+	b := r.URL.Query().Get(parBlock)
 	var timeout int
 	if b != "" {
 		var err error
 		timeout, err = strconv.Atoi(b)
-		if err != nil {
+		if err != nil || timeout < 0 {
 			writeError(w, http.StatusBadRequest, "bad block value, must be number of seconds")
 			return
 		}
 	}
 	log.Debugf("api timeout: %d", timeout)
-	ifNoneMatch := r.Header.Get("If-None-Match")
-
-	if apidTokenManager.getBearerToken() != ifNoneMatch {
-		w.Write([]byte(apidTokenManager.getBearerToken()))
+	ifNoneMatch := r.Header.Get(parTag)
+	log.Debugf("ifNoneMatch: %s", ifNoneMatch)
+	if a.tokenMan.getBearerToken() != ifNoneMatch {
+		w.Write([]byte(a.tokenMan.getBearerToken()))
 		return
 	}
 
 	select {
-	case <-apidTokenManager.getTokenReadyChannel():
-		w.Write([]byte(apidTokenManager.getBearerToken()))
+	case <-a.tokenMan.getTokenReadyChannel():
+		w.Write([]byte(a.tokenMan.getBearerToken()))
```

Comment:
This line look same as line 56, how does this channel stuff work?

Reply:
https://tour.golang.org/concurrency/5
The `select` will block until the tokenReadyChannel or the timeout channel is triggered. The first channel is a buffered channel, which is filled each time a new token is retrieval.  When Gateway is already there doing long-polling and blocked by this `select`, a new retrieved token will trigger this `case <-a.tokenMan.getTokenReadyChannel()` and unblock the Gateway. In Go, `select` will evaluate the each `case`'s argument first, which means `a.tokenMan.getTokenReadyChannel()` is executed immediately when `select` is reached, and the returned channel (`tokenMan.tokenUpdatedChan` in this case) is used for blocking.
Please check token.go#213 to check the implementation details. The `tokenUpdatedChan` is a buffered channel, and token manager sends to the channel by a `select`-`default` clause, so token manager can block api manager, but not the vice versa.
As I commented in token.go#212 `TODO`, this design assumes apid-gateway is 1-1 mapping, which is true for Alpha scope. @f1erro may want to reconsider this and make use of generic long-polling provided by apid-core in the future.